apps/aecore/src/aec_pinning_agent.erl

```diff
@@ -8,120 +8,192 @@
+
+%%%=============================================================================
+%%% API
+%%%=============================================================================
+
+-spec start_link(term(), atom()) -> {ok, pid()} | {error, {already_started, pid()}} | ignore | {error, Reason::any()}.
+start_link(Contract, PinningBehavior) ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, [Contract, PinningBehavior], []).
+
+%%%=============================================================================
+%%% Gen Server Callbacks
+%%%=============================================================================
+
+init([Contract, PinningBehavior]) ->
+    case PinningBehavior of
+        true ->
+            State = #state{contract = Contract},
+            lager:debug("started pinning agent"),
+            aec_events:subscribe(new_epoch),
+            {ok, State};
+        false ->
+            lager:debug("default pinning off - no agent started", []),
+            ignore
+    end.
+
+
+handle_call(_Request, _From, LoopState) ->
+    Reply = ok,
+    {reply, Reply, LoopState}.
+
+handle_cast(_Msg, LoopState) ->
+    {noreply, LoopState}.
+
+handle_info({gproc_ps_event, new_epoch, #{info := EpochInfo}}, #state{contract = Contract} = State) ->
+    Reply = spawn_for_epoch(EpochInfo, Contract),
+    {Reply, State};
+handle_info(_Info, LoopState) ->
+    {noreply, LoopState}.
+
+terminate(_Reason, _LoopState) ->
+    ok.
 
+code_change(_OldVsn, LoopState, _Extra) ->
+    {ok, LoopState}.
+
+    % case aec_parent_connector:has_parent_account(LastLeader) of
+    %     true -> aec_pinning_agent:spawn_for_epoch(NextEpochInfo, get_contract_pubkey(?ELECTION_CONTRACT), LastLeader);
+    %     false -> lager:debug("AGENT: No parent chain account found for ~p", [LastLeader])
+    % end
+stop() ->
+    lager:debug("STOPPED"),
+    gen_server:stop(?SERVER).
 %%%=============================================================================
 %%% API
 %%%=============================================================================
 
 
+%%%=============================================================================
+%%% Pinning Worker Process
+%%%=============================================================================
+
+spawn_for_epoch(#{last := Last} = EpochInfo, Contract) ->
+    {ok, LastLeader} = aec_consensus_hc:leader_for_height(Last),
+    case aec_parent_connector:has_parent_account(LastLeader) of
+        true -> spawn_for_epoch(EpochInfo, Contract, LastLeader);
+        false -> lager:debug("no pin in epoch, no parent account for ~p", [LastLeader]), noreply
+    end.
+
 spawn_for_epoch(EpochInfo, Contract, LastLeader) ->
-    lager:debug("AGENT: Trying to spawn pinning agent...", []),
     try
     case whereis(my_unique_process) of
         undefined ->
             Pid = spawn(aec_pinning_agent, start, [EpochInfo, Contract, LastLeader]),
-            register(?MODULE, Pid),
-            Pid;
+            register(?WORKER, Pid),
```

Comment:
This is not the unique name `my_unique_process` that is tested for above?  

But bigger question, why on earth do we spawn yet another process in here, and also wait for another top change.